webapp/channels/src/components/autosize_textarea.tsx

```diff
@@ -13,203 +14,169 @@ type Props = {
     onWidthChange?: (width: number) => void;
     onInput?: (e: FormEvent<HTMLTextAreaElement>) => void;
     placeholder?: string;
-    forwardedRef?: ((instance: HTMLTextAreaElement | null) => void) | React.MutableRefObject<HTMLTextAreaElement | null> | null;
 }
 
-export class AutosizeTextarea extends React.PureComponent<Props> {
-    private height: number;
-
-    private textarea?: HTMLTextAreaElement;
-    private referenceRef: React.RefObject<HTMLTextAreaElement>;
-    private measuringRef: React.RefObject<HTMLDivElement>;
-
-    constructor(props: Props) {
-        super(props);
-
-        this.height = 0;
-
-        this.referenceRef = React.createRef();
-        this.measuringRef = React.createRef();
-    }
-
-    componentDidMount() {
-        this.recalculateHeight();
-        this.recalculateWidth();
-    }
-
-    componentDidUpdate() {
-        this.recalculateHeight();
-        this.recalculateWidth();
-        this.recalculatePadding();
-    }
+const styles: { [Key: string]: CSSProperties} = {
+    container: {height: 0, overflow: 'hidden'},
+    reference: {height: 'auto', width: '100%'},
+    placeholder: {overflow: 'hidden', textOverflow: 'ellipsis', opacity: 0.5, pointerEvents: 'none', position: 'absolute', whiteSpace: 'nowrap', background: 'none', borderColor: 'transparent'},
+    measuring: {width: 'auto', display: 'inline-block'},
+};
 
-    private recalculateHeight = () => {
-        if (!this.referenceRef.current || !this.textarea) {
+const AutosizeTextarea = React.forwardRef<HTMLTextAreaElement, Props>(({
+    // TODO: The provided `id` is sometimes hard-coded and used to interface with the
+    // component, e.g. `post_textbox`, so it can't be changed. This would ideally be
+    // abstracted to avoid passing in an `id` prop at all, but we intentionally maintain
+    // the old behaviour to address ABC-213.
+    id = 'autosize_textarea',
+    disabled,
+    value,
+    defaultValue,
+    onChange,
+    onHeightChange,
+    onWidthChange,
+    onInput,
+    placeholder,
+}: Props, ref) => {
```

Comment:
We expand the props in the argument list to make it easier to track unused props.